### PR TITLE
LibWeb: Some row span fixes when combined with nested tables

### PR DESCRIPTION
**Pull request by axgallo**
1. Add logic to compute {min, max}-height.
2. Use min-height when calculating table height, per specifications.
3. Fix table width algorithm when available space is a constraint.

Fixes some issues with phylogenetic tree visualizations on Wikipedia.

### CODE REVIEW

**Comment by axgallo** [comment]
Before:
before

After:
after

File: Userland/Libraries/LibWeb/Layout/TableFormattingContext.h [hunk 1]
```diff
@@ -61,12 +62,22 @@ class TableFormattingContext final : public FormattingContext {
         double percentage_width { 0 };
     };
 
+    enum class RowType {
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/Layout/TableFormattingContext.h:65]: since this exactly matches `ColumnType` maybe we can reuse single enum definition for both rows and column?

Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/TableFormattingContext.h:65]: It is tempting, otoh they're still different things and wanted them spelled out explicitly. Maybe I can do some type aliases for that.

Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/TableFormattingContext.h]: Actually it's a size in both cases after all, fixed.

File: Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp [hunk 2]
```diff
@@ -468,6 +468,118 @@ void TableFormattingContext::determine_intrisic_size_of_table_container(Availabl
     }
 }
 
+void TableFormattingContext::compute_row_measures(AvailableSpace const& available_space)
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp:471]: could we replace `compute_column_measures` and `compute_row_measures` with one function that calculate measures depending on specified direction?

File: Userland/Libraries/LibWeb/Layout/LayoutState.cpp [hunk 3]
```diff
@@ -354,9 +354,9 @@ AvailableSpace LayoutState::UsedValues::available_inner_space_or_constraints_fro
     auto inner_width = available_width_inside();
     auto inner_height = available_height_inside();
 
-    if (inner_width.is_indefinite() && outer_space.width.is_intrinsic_sizing_constraint())
+    if ((inner_width.is_indefinite() || !isfinite(inner_width.to_px().value())) && outer_space.width.is_intrinsic_sizing_constraint())
         inner_width = outer_space.width;
-    if (inner_height.is_indefinite() && outer_space.height.is_intrinsic_sizing_constraint())
+    if ((inner_height.is_indefinite() || !isfinite(inner_height.to_px().value())) && outer_space.height.is_intrinsic_sizing_constraint())
         inner_height = outer_space.height;
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/Layout/LayoutState.cpp:360]: since this change affects all formatting contexts I suggest to move this change into separate commit with description why this additional check is needed.

File: Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp (hunk 2 above)
Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp:471]: I've tried it, with templates. Unfortunately, it requires around 7-8 helpers to abstract differences away, in the end it didn't reduce the line count and didn't look like a clear win. I'm agnostic about it, if you really want it that way I can rework it.

Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp:471]: I went a simpler way in GridFormattingContext where parameter that indicates whether rows and columns are processed is passed as a regular argument. you could try that too if you want :)

**Comment by kalenikaliaksandr** [comment]
Commit message "LibWeb: Some row span fixes when combined with nested tables" doesn't really seem to tell what have been changes. I see more detailed list of changes in commit description but maybe you should do a commit per item from this list then? :)

File: Userland/Libraries/LibWeb/Layout/LayoutState.cpp (hunk 3 above)
Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/LayoutState.cpp:360]: Done. I'm not sure if it currently affects any other formatting context, it requires a pretty particular set of circumstances (nested tables, basically).

File: Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp (hunk 2 above)
Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp]: The main problem is the multiplexing of logic, not how we route it. I've unified them, in any case.

**Comment by axgallo** [comment]
> Commit message "LibWeb: Some row span fixes when combined with nested tables" doesn't really seem to tell what have been changes. I see more detailed list of changes in commit description but maybe you should do a commit per item from this list then? :)

They're split now.

File: Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp [hunk 4]
```diff
@@ -152,107 +152,177 @@ void TableFormattingContext::calculate_row_column_grid(Box const& box)
     m_columns.resize(x_width);
 }
 
-void TableFormattingContext::compute_table_measures()
+void TableFormattingContext::compute_table_measures(TableDimension dimension, AvailableSpace const& available_space)
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp:155]: can we separate part of `compute_table_measures` that does measurements for cells and leave it to just take care of rows/columns depending on passed dimension?

Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp]: Done, I think.

Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp]: it should be possible to calculate all metrics for table cells in one pass so we don't need to call it twice for rows and columns separately.

Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp]: Done.

File: Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp [hunk 5]
```diff
@@ -152,107 +152,173 @@ void TableFormattingContext::calculate_row_column_grid(Box const& box)
...
+        cell.min_width = min_width + cell_intrinsic_width_offsets;
+        cell.max_width = max(max(width, min_width), max_width) + cell_intrinsic_width_offsets;
     }
+}
 
-    for (auto& cell : m_cells) {
-        if (cell.column_span == 1) {
-            m_columns[cell.column_index].min_width = max(m_columns[cell.column_index].min_width, cell.min_width);
-            m_columns[cell.column_index].max_width = max(m_columns[cell.column_index].max_width, cell.max_width);
+void TableFormattingContext::compute_table_measures(TableDimension dimension)
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp:217]: okay, after separation of cells measurements into `compute_cell_measures` I need to admit that `compute_table_measures` doesn't look neat with a bunch of `if (is_row_dim)`. probably in the future we would have to either separate those into different methods or find a way to have the same type to store rows and columns to avoid is_row_dim checks all over the place.

File: Userland/Libraries/LibWeb/Layout/LayoutState.cpp [hunk 6]
```diff
@@ -354,9 +354,9 @@ AvailableSpace LayoutState::UsedValues::available_inner_space_or_constraints_fro
     auto inner_width = available_width_inside();
     auto inner_height = available_height_inside();
 
-    if (inner_width.is_indefinite() && outer_space.width.is_intrinsic_sizing_constraint())
+    if ((inner_width.is_indefinite() || !isfinite(inner_width.to_px().value())) && outer_space.width.is_intrinsic_sizing_constraint())
         inner_width = outer_space.width;
-    if (inner_height.is_indefinite() && outer_space.height.is_intrinsic_sizing_constraint())
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/Layout/LayoutState.cpp:359]: this check means that somewhere we have a call to construct `AvailableSize` using `make_definite(INFINITY)` which is wrong and `make_indefinite()`, `make_min_content()` or `make_max_content()` should have been used instead.

I tried to add verification that value passed in `AvailableSize::make_definite` is actually definite:
```diff
diff --git a/Userland/Libraries/LibWeb/Layout/AvailableSpace.cpp b/Userland/Libraries/LibWeb/Layout/AvailableSpace.cpp
index 3cd4004f84..1622d5db04 100644
--- a/Userland/Libraries/LibWeb/Layout/AvailableSpace.cpp
+++ b/Userland/Libraries/LibWeb/Layout/AvailableSpace.cpp
@@ -11,6 +11,7 @@ namespace Web::Layout {

 AvailableSize AvailableSize::make_definite(CSSPixels value)
 {
+    VERIFY(isfinite(value.value()));
     return AvailableSize { Type::Definite, value };
 }
 ```
 
 and it showed up that `Tests/LibWeb/Layout/input/table/in-auto-height-flex-item.html` currently constructs invalid `AvailableSize`. so now you have a reduced example to fix root cause instead of introducing a new hack! :)

File: Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp (hunk 5 above)
Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp:217]: That's fair, at least we prevent divergence of behavior between dimensions for now. We should be able to sort this out in a future PR if we rework the row and column structures a bit, indeed.

File: Userland/Libraries/LibWeb/Layout/LayoutState.cpp (hunk 6 above)
Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/LayoutState.cpp]: I made `compute_table_box_width_inside_table_wrapper` a bit more careful with constraints vs definite sizes, which should address your concern. There are a dozen more places where we lean on the actual value of `min-content` and `max-content` which we might want to address in the future if we want to tighten the checks more.

Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/Layout/LayoutState.cpp]: thank you!

File: Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp (hunk 5 above)
Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/TableFormattingContext.cpp]: I've updated it to use templates and avoid bifurcation of logic based on `is_row_dim`. For `Cell` we need accessors to switch between retrieving row / column indices and spans based on the direction and I can't think of a way around that. For `Row` and `Column` it suffices to rename `{min, max}_width` and `{min, max}_height` to `{min, max}_size` to enable meta-programming, which I think is fine vs. forcing them into the same structure.

It's more code overall but the core algorithm doesn't have repetition or bifurcation based on dimension, so I'd go with it.

**Review by kalenikaliaksandr — APPROVED** [review]
